torch/_tensor.py

```diff
@@ -1012,6 +1013,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             else:
                 return _convert(ret, cls)
 
+    def __dlpack__(self, stream=None):
+        """
+        Creates a dlpack capsule of the current tensor to
+        be exported to other libraries.
+
+        This function will be called from the `from_dlpack` method
+        of the consumer library.
+
+        Args:
+            stream (integer or None): A Python integer representing a pointer
+            to a stream (CUDA or ROCm). `stream` is provided by the consumer
+            to the producer to instruct the producer to ensure that operations
+            can safely be performed on the array.
+            The pointer must be a positive integer or
+            -1 . If stream is -1 , the value may be used by the consumer to
+            signal "producer must not perform any synchronization. Optional.
+        """
+        if has_torch_function_unary(self):
+            return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
+        if stream is not None and type(stream) is not int:
+            # Stream pointers in CUDA/ROCm are uniquely numbered and can
+            # be retrieved from their integer value.
+            raise TypeError('stream must be ``int`` or ``none``')
+        elif stream is not None and stream != -1:
+            if self.device.type == 'cuda':
+                stream = torch.cuda.streams.ExternalStream(stream)
+                # Only synchronize on different streams
+                if stream != torch.cuda.current_stream:
+                    event = torch.cuda.Event()
+                    event.record(stream)
+                    torch.cuda.current_stream().wait_event(event)
+        return torch.to_dlpack(self)
+
+    def __dlpack_device__(self) -> Tuple[enum.IntEnum, int]:
+        class DLPackIds(enum.IntEnum):
+            cpu = 1
+            cpu_pinned = 3
+            cuda = 2
+            opencl = 4
+            vulkan = 7
+            rocm = 10
+
+        if has_torch_function_unary(self):
+            return handle_torch_function(Tensor.__dlpack_device__, (self,), self)
+        idx = self.device.index if self.device.index is not None else 0
+        device_type = self.device.type
+        if device_type == 'cuda' and torch.version.hip is not None:
+            device_type = 'rocm'
+        elif device_type == 'cpu' and self.is_pinned():
+            device_type = 'cpu_pinned'
```

Comment:
Question: `device_type` will not be `'cpu'` in PyTorch?

Reply:
sorry I dont' follow you

Reply:
My bad, sorry Emilio. I was being silly early on...